refactoring

majkinetor opened the pull request:
I changed the vagrant so that no changes to it are necessary.

I moved inline script to `scripts` folder and named it `TestPackage`. It now contains logic to look for nupkg files in the c:\packages folder for local source and c:\packages\remote for remote source.

So working with this now requires only to copy files to packages without any other action.

ferventcoder commented:
This assumes one package in the folder. I will need to think on this a bit as it is not typically how I use the environment (although many others probably would).

ferventcoder commented on shell/TestPackage.ps1:
```diff
@@ -0,0 +1,24 @@
+# https://github.com/chocolatey/choco/issues/512
+$validExitCodes = @(0, 1605, 1614, 1641, 3010)
+
+$package  = ls c:\packages\*.nupkg
```
rename to `$packages`

majkinetor replied:
Then the next one will be wrongly named, the one that takes just the first one :)

ferventcoder replied:
I may have missed that - I saw you split the name of the file, but I may have missed where you chose only the first package in the directory.

majkinetor replied:
Or I can make it run ALL packages from there, 1 by 1.

ferventcoder replied:
Right, kind of what I was thinking.

majkinetor replied:
I wanted to do that, but then, why would you run more then 1 package ? It kinds defits the purpose of clean environment for one, and second, you can always add more while machine is running.

ferventcoder replied:
Maybe you are testing other things. Who knows? I typically don't like changes that limit the possibility further. More thinking usually brings the best option available to meet all needs.

ferventcoder commented on shell/TestPackage.ps1:
```diff
@@ -0,0 +1,24 @@
+
+$package  = ls c:\packages\*.nupkg
+if (!$package) { $package = ls c:\packages\remote\*.nupkg; $remote=$true }
+if (!$package) { Write-Host 'No packages found in directory packages or packages\remote'; return }
```
Not really sure why a need to have two directories. Is it confusing to always have the secondary source here?

majkinetor replied:
How would you handle the difference ? How would you mark whats needed for remote and what with local source ?

ferventcoder replied:
It already does this.

majkinetor replied:
So why did you have 2 choco install lines there and asking to choose one ? I added remote just because it was there.

majkinetor commented:
> This assumes one package in the folder. I will need to think on this a bit as it is not typically how I use the environment (although many others probably would).

So how do you use it ? I planned to integrate it with AU's `Test-Package` so that when you call something like `Test-Package -Vagrant` it will copy the files there and run the machine.

But anway, just say how u use it so I can make it work for all of us.

ferventcoder commented:
> But anway, just say how u use it so I can make it work for all of us.

That's why the think on this...

majkinetor commented:
> That's why the think on this...

You didn't answer me how you use it. If you do, maybe we can both think about solution ?

ferventcoder commented:
> You didn't answer me how you use it. If you do, maybe we can both think about solution ?

In lots of ways. 

* Sometimes to test a simple package for folks (remote package, no copy local first). 
* Also for testing new Chocolatey releases/betas (keep in mind the packages folder may have a ton of different versions of the same package in it for me).
* Testing different scenarios with Chocolatey.


I like that the Vagrantfile allows for me to specify as few or as many files to test as I want.

ferventcoder commented:
> I planned to integrate it with AU's `Test-Package` so that when you call something like `Test-Package -Vagrant` it will copy the files there and run the machine.

We can definitely get it set up for that scenario as well. Perhaps a new folder where it will just look and install the packages in that folder automatically.

majkinetor commented:
> Sometimes to test a simple package for folks (remote package, no copy local first).

I see, we could solve that with either vagrant argument or env var that contains package name, in which case c:\packages folder wouldn't get read.

So, one of those 2

      $env:package = 'copyq'; vagrant up --provision
      vagrant up --provision --package copyq

TBH, I think the second one is not as easy to implement (at least it wasn't when I was a lot into vagrant a year before) but its doable. I think first one is probably the best.

> Also for testing new Chocolatey releases/betas (keep in mind the packages folder may have a ton of different versions of the same package in it for me). 
> ... 
> I like that the Vagrantfile allows for me to specify as few or as many files to test as I want.

Installing each package is a trivial change I can add it.

majkinetor commented:
Yeah, this one also: 

      $env:package = 'copyq dbeaver eac'; vagrant up --provision

ferventcoder commented:
Here's the path I would suggest going down for now:

Add another folder - use the PowerShell script to run that if there are packages.

Leave the vagrantfile changes alone, only add in the additional script call.

majkinetor commented:
Are you sure you want that? It doesn't seem clean too me, multiple different paths doing the installs, not DRY and somewhat confusing. Why would you ever change vagrantfile ? Everything should be achievable via env vars.

majkinetor commented:
There is only one benefit in doing os - its persisted. But, you can equally persist env vars in vagrant file itself on the very top of the file (more visible). I could also add something like `/shell/User.ps1` that will let user add its script to be executed at the end, I think way better then what we have now with inline script where you must be careful with escapes and its harder to debug.

ferventcoder commented:
> Are you sure you want that? It doesn't seem clean too me, multiple different paths doing the installs, not DRY and somewhat confusing. Why would you ever change vagrantfile ? Everything should be achievable via env vars.

You are wanting to support different ways of working with the testing environment. Taking out what you want to achieve now and going to the original stated use of this module, what is wrong with it in its current form?

majkinetor commented:
OK, I worked out on some things we talked about.

- I refactored vagrantfile to get DRY
- Added User.ps1 script in repository root (for visibility) that contains code from insline script
- Updated TestPackages script to allow multiple remote + local packages with given versions

Example of use: `$Env:PACKAGES='copyq dbeaver:2.7.1'; vagrant up ...`

This will take all packages in c:\packages (remote is removed) and remote packages given in env var. and install them. And if you do nothing as before those changes it will behave the same, you just don't edit `Vagrantfile` but `User.ps1` which is IMO much better.

Anyway, I just wanted you to see my changes, perhaps you want to incorporate something or everything or nothing at all.

majkinetor commented:
@ferventcoder @gep13 

Integrated AU `Test-Package` with my fork.  It can now test install, uninstall or both, along with parameters. 

Vagrant environment path is specified via parameter `Vagrant` or via env. var `$au_Vagrant`. I currently keep in my profile `$Env:au_Vagrant = 'c:\chocolatey\chocolatey-test-environment'`. Parameter `VagrantClear` will remove all packages from `packages` folder prior to testing. 

Example:

```Powershell
C:\work\au-packages\yed> Test-Package  -VagrantClear -Parameters '/Shortcut'

Package info
  Path:         C:\work\au-packages\yed\yed.3.16.2.1.nupkg
  Name:         yed
  Version:      3.16.2.1
  Parameters:   /Shortcut
  Vagrant:      c:\work\_chocolatey\chocolatey-test-environment

Testing package using vagrant
Removing existing vagrant packages
<starts vagrant in another shell>
```

In another shell

```
==> default: Running provisioner: shell...
    default: Running: shell/TestPackages.ps1 as c:\tmp\vagrant-shell.ps1
==> default: ============================================================
==> default: TESTING FOLLOWING PACKAGES: yed:3.16.2.1
==> default: ============================================================
==> default:
==> default: ------------------------------------------------------------
==> default: PACKAGE: yed:3.16.2.1
==> default:
==> default: OPTIONS: c:\packages\yed.3.16.2.1.xml
==> default:
==> default: Name                           Value
==> default: ----                           -----
==> default: Install                        True
==> default: Uninstall                      True
==> default: Parameters                     /Shortcut
==> default: ------------------------------------------------------------
==> default:
==> default: TESTING INSTALL FOR yed:3.16.2.1
==> default: Choco cmd: choco install -fy yed --allow-downgrade --version 3.16.2.1 --source "'c:\packages;http://chocolatey.org/api/v2/'"  --params System.Collections.Hashtable.Parameters
==> default: Chocolatey v0.10.3
==> default: Installing the following packages:
==> default:
==> default: yed
==> default: By installing you accept licenses for the packages.
==> default: yed v3.16.2.1 (forced)
==> default: yed package files install completed. Performing other installation steps.
==> default: File appears to be downloaded already. Verifying with package checksum to determine if it needs to be redownloaded.
==> default: Hashes match.
==> default: Hashes match.
==> default:
==> default: Extracting C:\Users\Administrator\AppData\Local\Temp\chocolatey\yed\3.16.2.1\yEd-3.16.2.1.zip to C:\ProgramData\chocolatey\lib\yed\tools...
==> default: C:\ProgramData\chocolatey\lib\yed\tools
==> default:
==> default: Yed installed in: C:\ProgramData\chocolatey\lib\yed\tools\yed-3.16.2.1
==> default: yed.cmd added to chocolatey bin directory
==> default:  The install of yed was successful.
==> default:
==> default:   Software installed to 'C:\ProgramData\chocolatey\lib\yed\tools'
==> default:
==> default: Chocolatey installed 1/1 packages. 0 packages failed.
==> default:  See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
==> default:
==> default: Did you know the proceeds of Pro (and some proceeds from other
==> default:  licensed editions) go into bettering the community infrastructure?
==> default:  Your support ensures an active community, it makes you look smarter,
==> default:  plus it nets you some awesome features!
==> default:  https://chocolatey.org/compare
==> default: Exit code for yed:3.16.2.1 was 0
==> default:
==> default: TESTING UNINSTALL FOR yed:3.16.2.1
==> default: Choco cmd: choco uninstall -fy yed
==> default: Chocolatey v0.10.3
==> default: Uninstalling the following packages:
==> default:
==> default: yed
==> default: yed v3.16.2.1 (forced)
==> default:  Skipping auto uninstaller - No registry snapshot.
==> default:  yed has been successfully uninstalled.
==> default: Chocolatey uninstalled 1/1 packages. 0 packages failed.
==> default:  See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
==> default: Running provisioner: shell...
    default: Running: User.ps1 as c:\tmp\vagrant-shell.ps1
==> default: User script can be added in User.ps1
==> default:
==> default: SUCCESS: Specified value was saved.
==> default: Testing package if a line is uncommented.
==> default: Exit code was 0
```

majkinetor commented:
@ferventcoder 

OK, I will close this as I can no longer maintain my master branch and this branch. 

All changes are now in my master branch so if/when you want me I can PR it. 

AU `Test-Package` now links to this environment as it doesn't work with original. I did few other changes that significantly speed things up - now package testing provisioners run each time and infrastructure setting runs only once as before; this makes testing much faster and you litterally have to do nothing in test environment except clone it, all other things are done with `Test-Package`. I might add more machines in the future via parameter. 
 
https://github.com/majkinetor/chocolatey-test-environment
